### PR TITLE
✨ Add way for users to generate commit messages with AI

### PR DESCRIPTION
**Pull request by segersniels**
## Description

This PR introduces a new subcommand `generate` which allows the user to generate a commit message using OpenAI's GPT model. It keeps generating until the user confirms and then it gets passed to `withClient` to actually commit.

It currently passes an empty `message` body so not sure if we still want to prompt the user for a body after generating the title. It would also need to potentially respect the new `MESSAGE_PROMPT` config from #1072.

Issue: #1069

## Tests

- [x] All tests passed.

### CODE REVIEW

**Comment by segersniels** [comment]
Still needs a couple of tweaks. Like properly checking the total amount of tokens used and adjusting the max tokens available depending on the model (hardcoded to 3.5 currently).

**Comment by segersniels** [comment]
@carloscuesta This one is now also ready to be reviewed. Go ham on what needs to change.

**Comment by segersniels** [comment]
Screen Recording 2023-05-06 at 20 21 41

**Comment by segersniels** [comment]
The generated prompt sent to OpenAI's API is something that needs to be experimented with and improved through experimentation. This is what I came up with for my implementation at https://github.com/segersniels/gitmoji but it's definitely not perfect yet.

**Comment by segersniels** [comment]
Moving the to be used model to the config might be interesting as well since it's not great UX if users need to constantly add `--model gpt-4`. Only reason why I didn't is that it feels out of place for users that will never use these AI generated messages.

**Comment by carloscuesta** [comment]
Hey! @segersniels 

Thanks for raising the PR so fast! Feel free to rebase it now as everything else has been merged! I'll review it once you rebase it 🙏🏼 😊

**Comment by segersniels** [comment]
> Hey! @segersniels
> 
> Thanks for raising the PR so fast! Feel free to rebase it now as everything else has been merged! I'll review it once you rebase it 🙏🏼 😊
> 
> I'll do a release right now to ship your changes 🚀

Great to hear, everything should be rebased. Probably needs some more coverage testing but the base should be ready to review.

File: src/cli.js [hunk 1]
```diff
@@ -35,6 +36,7 @@ const cli = meow(
     remove          Remove a previously initialized commit hook
     search          Search gitmojis
     update          Sync emoji list with the repo
+    generate        Generate a commit message using OpenAI's GPT model
```
Comment (carloscuesta) [inline src/cli.js:39]: How would that work for people using the hook? 🤔

File: src/commands/generate/index.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import withClient from '@commands/commit/withClient'
+
+type CommitOptions = {
+  model?: 'gpt-3.5-turbo' | 'gpt-4',
+  context?: boolean
+}
+
+const generate = async (options: CommitOptions): Promise<void> => {
+  if (!process.env.OPENAI_API_KEY) {
+    return console.error(
```
Comment (carloscuesta) [inline src/commands/generate/index.js:17]: Let's cover this with a test!

File: src/commands/generate/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      chalk.red(`Unable to locate OPENAI_API_KEY in environment`)
+    )
+  }
+
+  const { stdout: diff } = await execa('git', ['diff', '--cached'])
+  if (!diff) {
+    return console.error(chalk.red('Could not find any staged files to commit'))
```
Comment (carloscuesta) [inline src/commands/generate/index.js:24]: 👀 Test here also

File: src/commands/generate/index.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  // Let the user specify additional context to give some guidance to the generated message
+  let context
```
Comment (carloscuesta) [inline src/commands/generate/index.js:28]: Can we define this as a const? 🙏🏼

File: src/utils/findGitmojiCommand.js [hunk 5]
```diff
@@ -7,18 +7,23 @@ const isSupportedCommand = (command: ?string, options: Object): boolean => {
...
+    case FLAGS.COMMIT:
+    case FLAGS.HOOK:
+      return {
+        message: flags['message'],
+        mode: command === FLAGS.HOOK ? COMMIT_MODES.HOOK : COMMIT_MODES.CLIENT,
+        scope: flags['scope'],
+        title: flags['title']
+      }
+    case FLAGS.GENERATE:
+      return {
```
Comment (carloscuesta) [inline src/utils/findGitmojiCommand.js:20]: We need to cover this with a test!

**Review by carloscuesta — COMMENTED** [review]
I'll continue my review later!! 

Thanks ❤️

**Comment by segersniels** [comment]
Closing for now until more people ask for it. If people really want to use something similar there is already a mention to `genmoji` in the related tools on the website.

**Pull request closed without merging**